README.md
mongoose-encrypt
================

Transparent encryption for Mongoose fields with built-in password migration.

Allows you to easily encrypt `String` fields using `aes-256-cbc`.


Idea
====

Store a timestamp along the encrypted data to allow painless password migration.

The data that is stored in MongoDB as base64 encoded strings and consists of (from left to right)

* the string `ENCRYPTED___` to handle a mix of encrypted/unencrypted data (**you can drop this plugin into your existing data**),
* 8 chars representing the seconds since the Unix epoch in hex (for range query pleasure and password migration),
* a 16 chars (8 byte) salt hex string which is randomly generated for every encrypted string and appended to the password before encrypting and decrypting,
* and the encrypted data itself as base64 string.


Usage
=====

First `npm install mongoose-encrypt`.

Now imagine a website where users sign up with their Twitter account. It's probably a good idea to encrypt the OAuth token.

```js
var encrypt = require('mongoose-encrypt');

var userSchema = new Schema({
	createdAt: Date,
	twitter: {
		name: String,
		token: String
	}
});

userSchema.plugin(encrypt, {
	paths: ['twitter.token'],
	password: function(date) {
		//Return the correct password for the given date.
		//As long as you don't need to migrate to a new password, just return the current one.
		return process.env.AES_ENCRYPTION_PASSWORD;
	}
});
```

That's it! The plugin sets up a `getter` and `setter` to decrypt and encrypt each path on the fly using `aes-256-cbc`.


Use cases
=========

As mentioned above storing OAuth tokens or similar in plain text is probably a bad idea. Additional this plugin was created to securely store bank account data on behalf of users.


Heads up
========

I'm not a security expert. Not at all. If you have any concerns regarding this plugin please create an issue (or contact me via e-mail if it's a critical issue).

Also:

* It's a good idea to not store or hardcode the encryption key (the example uses an environment variable)
* This plugin will only secure your data in case someone gets access directly to your database (physically or otherwise)
* You still need to make sure the data is transmitted securely (e.g. using TLS)
* If someone gets access to your application server (not just the database), you're screwed anyway


HISTORY.md
0.0.1-alpha.2 (2014-05-12)
--------------------------

* Removed unused `async` dependency.

test/index.js
var assert = require('assert');
var mongoose = require('mongoose');
var Schema = mongoose.Schema;
var encrypt = require('../lib');

mongoose.connect('mongodb://localhost/mongoose-encrypt');
mongoose.set('debug', false);

var userSchema = new Schema({
	_id: String,
	secret1: String,
	nested: {
		secret2: String
	}
});

//We'll later adjust this date to simulate password migration.
var newPasswordDate = new Date(2070, 01, 01);

userSchema.plugin(encrypt, {
	paths: ['secret1', 'nested.secret2'],
	password: function(date) {
		if(date >= newPasswordDate) {
			return 'correcthorsebatterystaple';
		} else {
			return 'keyboardkitten';
		}
	}
});

var User = mongoose.model('User', userSchema);

describe('Paths which are not of type String', function() {
	it('should throw an error', function(done) {
		assert.throws(function() {
			(new Schema({
				not_a_string: Number
			})).plugin(encrypt, {
				paths: 'not_a_string',
				password: Math.random()
			});
		});

		done();
	});
});

describe('Basic encryption', function() {
	before(function(done) {
		//Clear the DB.
		User.remove({}, done);
	});

	it('should successfully save a model', function(done) {
		var user = new User({
			_id: 'user1',
			secret1: 'secret'
		});

		user.save(done);
	});

	it('should return an encrypted value when bypassing the plugin by using lean()', function(done) {
		User.findOne({_id: 'user1'}).lean().exec(function(err, user) {
			assert.ifError(err);

			assert.notEqual(user.secret1, 'secret', 'The secret1 is not in plain text.');
			assert(/^ENCRYPTED___/.test(user.secret1), 'The secret1 starts with the magic prefix.');

			assert.strictEqual(user.nested, undefined, 'The nested.secret2 is not set.');

			done();
		});
	});

	it('should decrypt the value when using Mongoose as usual', function(done) {
		User.findOne({_id: 'user1'}).exec(function(err, user) {
			assert.ifError(err);

			assert.strictEqual(user.secret1, 'secret', 'The secret1 is readable.');
			assert.strictEqual(user.nested.secret2, undefined, 'The nested.secret2 is not set.');

			done();
		});
	});

	it('should handle value updates by encrypting the new value', function(done) {
		User.findOne({_id: 'user1'}).exec(function(err, user) {
			assert.ifError(err);

			user.secret1 = 'new secret';
			user.nested.secret2 = 'new secret 2';

			user.save(function(err) {
				assert.ifError(err);

				User.findOne({_id: 'user1'}).exec(function(err, user) {
					assert.ifError(err);

					assert.strictEqual(user.secret1, 'new secret', 'The secret1 has been updated.');
					assert.strictEqual(user.nested.secret2, 'new secret 2', 'The nested.secret2 has been updated.');

					done();
				});
			});
		});
	});

	it('should not double encrypt a value that is already encrypted', function(done) {
		// Encrypted value can get passed around accidentally if it's nested
		var user2 = new User({
			_id: 'user2',
			nested: {
				secret2: 'new secret 2'
			}
		});
		User.findOne({_id: 'user1'}).exec(function(err, user) {
			assert.ifError(err);

			user.nested = user2.nested;

			user.save(function(err) {
				assert.ifError(err);

				User.findOne({_id: 'user1'}).exec(function(err, user) {
					assert.ifError(err);
					assert.strictEqual(user.nested.secret2, 'new secret 2', 'The nested.secret2 has been updated.');
					done();
				});
			});
		});
	});

	it('should leave unset properties untouched', function(done) {
		//Only fetch the _id property.
		User.findOne({_id: 'user1'}, {_id: 1}).exec(function(err, user) {
			assert.ifError(err);

			assert.strictEqual(user.secret1, undefined, 'The secret1 is not set.');
			assert.strictEqual(user.nested.secret2, undefined, 'The nested.secret2 is not set.');

			done();
		});
	});
});

describe('Legacy data, which is already present and unencrypted', function() {
	before(function(done) {
		var user = {
			_id: 'user2',
			secret1: 'secret'
		};

		//Save a user to the DB, but bypass Mongoose.
		User.collection.insert(user, done);
	});

	it('should return the unencrypted value when bypassing the plugin by using lean()', function(done) {
		User.findOne({_id: 'user2'}).lean().exec(function(err, user) {
			assert.ifError(err);

			assert.strictEqual(user.secret1, 'secret', 'The secret1 is readable in plain text.');
			assert.strictEqual(user.nested, undefined, 'The nested.secret2 is not set.');

			done();
		});
	});

	it('should not try to decrypt the legacy data and just return it', function(done) {
		User.findOne({_id: 'user2'}).exec(function(err, user) {
			assert.ifError(err);

			assert.strictEqual(user.secret1, 'secret', 'The secret1 is readable in plain text.');
			assert.strictEqual(user.nested.secret2, undefined, 'The nested.secret2 is not set.');

			done();
		});
	});

	it('should encrypt the data when saving the model back to the DB using Mongoose', function(done) {
		User.findOne({_id: 'user2'}).exec(function(err, user) {
			assert.ifError(err);

			user.secret1 = user.secret1;

			user.save(function(err) {
				assert.ifError(err);

				User.findOne({_id: 'user2'}).lean().exec(function(err, user) {
					assert.ifError(err);

					assert.notEqual(user.secret1, 'secret', 'The secret1 is not in plain text.');
					assert(/^ENCRYPTED___/.test(user.secret1), 'The secret1 starts with the magic prefix.');

					done();
				});
			});
		});
	});
});

describe('Password migration', function() {
	before(function(done) {
		newPasswordDate = new Date();

		//Wait a moment to make sure the password migration actually has to different dates to use.
		setTimeout(done, 10);
	});

	it('should use the old password for decryption and then the new one for encryption', function(done) {
		User.findOne({_id: 'user2'}).lean().exec(function(err, user) {
			assert.ifError(err);

			var oldEncryptedValue = user.secret1;
			assert.notEqual(user.secret1, 'secret', 'The secret1 is not in plain text.');

			User.findOne({_id: 'user2'}).exec(function(err, user) {
				assert.ifError(err);

				assert.strictEqual(user.secret1, 'secret', 'The secret1 is readable.');

				user.secret1 = user.secret1;

				user.save(function(err, user) {
					assert.ifError(err);

					//TODO: this one is failing, because when encrypting the data in `pre save` hook,
					//they are also encrypted when using this instance after saving.
					assert.strictEqual(user.secret1, 'secret', 'The secret1 is readable.');

					User.findOne({_id: 'user2'}).lean().exec(function(err, user) {
						assert.ifError(err);

						assert.notEqual(user.secret1, 'secret', 'The secret1 is not in plain text.');
						assert.notEqual(user.secret1, oldEncryptedValue, 'The two encrypted values differ.');

						done();
					});
				});
			});
		});
	});
});

lib/index.js
var crypto = require('crypto');

var MAGIC_PREFIX = 'ENCRYPTED___';//Aww Yiss, more underscores.
var CHECK_MAGIC_PREFIX = new RegExp('^' + MAGIC_PREFIX);
var ALGORITHM = 'aes-256-cbc';
var SALT_BYTE_LENGTH = 16;

//As a reminder, this is how we store the data (pipes not actually included).
//ENCRYPTED___|timestamp|salt|data

//The magic prefix begins at position 0 (duh).
var OFFSET_MAGIC_PREFIX = 0;
var LENGTH_MAGIC_PREFIX = MAGIC_PREFIX.length;

//The timestamp comes right after the magic prefix.
var OFFSET_TIMESTAMP = OFFSET_MAGIC_PREFIX + LENGTH_MAGIC_PREFIX;
var LENGTH_TIMESTAMP = 8;//4 bytes in hex characters.

//The salt comes right after the timestamp.
var OFFSET_SALT = OFFSET_TIMESTAMP + LENGTH_TIMESTAMP;
var LENGTH_SALT = SALT_BYTE_LENGTH * 2;//#hex = 2 * #bytes

//The data comes right after the timestamp and ends at the very end of the string.
var OFFSET_DATA = OFFSET_SALT + LENGTH_SALT;

var extractDate = function(rawData) {
	var hexTimestamp = rawData.substr(OFFSET_TIMESTAMP, LENGTH_TIMESTAMP);
	var timestamp = parseInt(hexTimestamp, 16);
	var date = new Date(timestamp * 1000);

	return hexTimestamp;
};

var extractSalt = function(rawData) {
	return rawData.substr(OFFSET_SALT, LENGTH_SALT);
};

var extractValue = function(rawData) {
	return rawData.substr(OFFSET_DATA);
};

var createHexTimestamp = function(unix) {
	//Left pad the hex timestamp with zeros.
	var hexTimestamp = ('00000000' + unix.toString(16)).substr(-8);

	return hexTimestamp;
};

var createSalt = function() {
	var salt = crypto.pseudoRandomBytes(SALT_BYTE_LENGTH).toString('hex');

	return salt;
};

var encrypt = function(data, password) {
	var cipher = crypto.createCipher(ALGORITHM, password);
	var encrypted = cipher.update(data, 'utf8', 'base64') + cipher.final('base64');

	return encrypted;
};

var decrypt = function(data, password) {
	var decipher = crypto.createDecipher(ALGORITHM, password);
	var decrypted = decipher.update(data, 'base64', 'utf8') + decipher.final('utf8');

	return decrypted;
};

module.exports = function(schema, options) {
	var paths = options.paths;
	var getPassword = options.password;

	var setter = function(rawValue) {
		if(!rawValue) {
			return rawValue;
		}

		var unix = Math.round(Date.now() / 1000);
		var now = new Date(unix);

		//Create a random salt.
		var salt = createSalt();

		//Get the password for the current date and time.
		var password = getPassword(now);

		var saltedPassword = password + salt;
		var encryptedValue = encrypt(rawValue, saltedPassword);

		//Compose the value we store.
		var value = MAGIC_PREFIX + createHexTimestamp(unix) + salt + encryptedValue;

		return value;
	};

	var getter = function(value) {
		if(!value) {
			return value;
		}

		//Nothing to do here, because the value is not encrypted.
		if(!CHECK_MAGIC_PREFIX.test(value)) {
			return value;
		}

		//Extract all the things that are encoded in the stored value.
		var date = extractDate(value);
		var password = getPassword(date);
		var salt = extractSalt(value);
		var encryptedValue = extractValue(value);

		var saltedPassword = password + salt;

		var rawValue = decrypt(encryptedValue, saltedPassword);

		return rawValue;
	};

	paths.forEach(function(path) {
		var pathMeta = schema.paths[path];

		//Make sure the path is of type String.
		if(!pathMeta || pathMeta.options.type !== String) {
			throw new Error('mongoose-encrypt can only be used on String properties. Was applied to "' + path + '" which is of type "' + pathMeta.options.type.name + '".');
		}

		schema.path(path).set(setter);
		schema.path(path).get(getter);
	});
};